### ISSUE TITLE
Support .yml as config file extension name

### ISSUE THREAD

**Issue by sghuang19** [DavidAnson/markdownlint-cli2#392]
Currently only .yaml extension is supported.

**Comment by DavidAnson** [comment]
The current behavior is deliberate in order to avoid supporting redundant file names - as with the `.markdownlint.*` set for legacy reasons. YAML is (slightly) more explicit than YML.

**Comment by sghuang19** [comment]
Hi David,

I'm still unclear why you think supporting different extensions is "redundant"? I can give two examples by ESLint (before flat config came out) and Prettier, they both support both `.yaml` and `.yml` extensions.

![image](https://github.com/user-attachments/assets/039a110e-4f5a-4104-b51f-caadf884b018)

![image](https://github.com/user-attachments/assets/d78b7eff-f955-49e3-b8fb-c4df14fd20b3)

A counter argument for myself is, YAML.org encourages using .yaml whenever possible.

https://yaml.org/faq.html

**Comment by DavidAnson** [comment]
I claim it's redundant to support the same file by two different names. It requires documenting both, looking for both on disk at runtime, resolving conflicts between them, providing editor support for both, etc.. Everything that's possible today with .yaml would still be possible with .yml - and nothing new. Because this file is unique to this tool and it has only ever supported .yaml, there shouldn't be any .yml files out there to be "backwards compatible" with.

**Comment by goto-dev-null** [comment]
I think this is a bizarre take. It's a bit like saying that YAML should not have the `yml` extension, period. But it does. So if you're going to support YAML, you should support both.

By this logic, why does this program accept jsonc, yaml, and cjs? Why not force people to use one?

> Everything that's possible today with .yaml would still be possible with .yml - and nothing new. 

Yes, something _would_ be new: it would allow markdownlint-cli2 to be consistent with _every other YAML file_ in our project.

You already have a method named [`processFirstMatchingConfigurationFile`](https://github.com/DavidAnson/markdownlint-cli2/blob/5387279b3b4c24822c0f86d4df4f28b37e3e8992/markdownlint-cli2.mjs#L288) that simply chooses files [in order of priority](https://github.com/DavidAnson/markdownlint-cli2/blob/5387279b3b4c24822c0f86d4df4f28b37e3e8992/markdownlint-cli2.mjs#L262-L265), so I don't understand what you mean about "looking for both on disk at runtime"; it already does that. And as for "resolving conflicts between them", you don't do that between the different supported filetypes, so why is yml different?

As for "documenting both", again, you've already documented jsonc, yaml, and cjs. You already have to document that both .markdownlint.yaml and .markdownlint.yml are supported; can you honestly say that doing so is _that_ much of an effort?

The level of effort for this would be trivial and, once established would take 0 upkeep, so it's really more an idealogical limitation than a technical one, and the limitation gives you nothing but costs you something (lack of standardization within a project). It's forcing a preference on people.

**Comment by DavidAnson** [comment]
> I think this is a bizarre take. … So if you're going to support YAML, you should support both.

Three-character extensions are a relic of DOS days and do not need to be carried forward indefinitely. The YAML organization itself shows a STRONG preference for “.yaml” with the short form seemingly used only for tooling that requires it. This seems compelling to me.

.yaml: 760 files
<https://github.com/search?q=org%3Ayaml+path%3A*.yaml&type=code>

.yml: 26 files
<https://github.com/search?q=org%3Ayaml+path%3A*.yml&type=code>

> By this logic, why does this program accept jsonc, yaml, and cjs? Why not force people to use one?

JSON and YAML are static and I would support only one of them if they weren’t roughly equally popular. JavaScript makes it possible to build configuration dynamically and is supported for scenarios where that is helpful.

> You already have a method named [[`processFirstMatchingConfigurationFile`](https://github.com/DavidAnson/markdownlint-cli2/blob/5387279b3b4c24822c0f86d4df4f28b37e3e8992/markdownlint-cli2.mjs#L288)](https://github.com/DavidAnson/markdownlint-cli2/blob/5387279b3b4c24822c0f86d4df4f28b37e3e8992/markdownlint-cli2.mjs#L288) that simply chooses files [[in order of priority](https://github.com/DavidAnson/markdownlint-cli2/blob/5387279b3b4c24822c0f86d4df4f28b37e3e8992/markdownlint-cli2.mjs#L262-L265)](https://github.com/DavidAnson/markdownlint-cli2/blob/5387279b3b4c24822c0f86d4df4f28b37e3e8992/markdownlint-cli2.mjs#L262-L265), so I don't understand what you mean about "looking for both on disk at runtime"; it already does that. …

Look again - there is a call to `fs.access` for every unique file name, and that function is invoked for every directory that is visited. I want to minimize the amount of work that happens there to make things run as quickly as I can.

> As for "documenting both", again, you've already documented jsonc, yaml, and cjs. You already have to document that both .markdownlint.yaml and .markdownlint.yml are supported; can you honestly say that doing so is _that_ much of an effort?
> 
> The level of effort for this would be trivial and, once established would take 0 upkeep, so it's really more an idealogical limitation than a technical one, and the limitation gives you nothing but costs you something (lack of standardization within a project). It's forcing a preference on people.

You are welcome to prioritize things differently in your own open source projects.

**Comment by goto-dev-null** [comment]
> The YAML organization itself shows a STRONG preference for “.yaml” with the short form seemingly used only for tooling that requires it. This seems compelling to me.

The keyword here is "preference".

> JSON and YAML are static and I would support only one of them if they weren’t roughly equally popular. 

So? Why care what's popular, if the intention is to just put forward your own preference?

> Look again - there is a call to fs.access for every unique file name, and that function is invoked for every directory that is visited. I want to minimize the amount of work that happens there to make things run as quickly as I can.

If this is true why are you invoking `fs.access` for each item in the list when you will only need to do that if it ends out being the last item? If you really cared about minimizing amount of work happening, you would only call `fs.access` for the second item if the first item is not present.

> You are welcome to prioritize things differently in your own open source projects.

You are of course welcome to prioritize things however you want in your own code, but "prioritization" isn't applicable here. It's preference. You can also enforce your own preference in your own code, but that should be the reason given, not saying it's "redundant" when you already do that same redundant behavior.

**Comment by DavidAnson** [comment]
> So? Why care what's popular, if the intention is to just put forward your own preference?

No, my point was that I would choose the one overwhelmingly popular choice - if such a thing existed. That would be not MY preference, but EVERYONE's preference.

> If this is true why are you invoking `fs.access` for each item in the list when you will only need to do that if it ends out being the last item? If you really cared about minimizing amount of work happening, you would only call `fs.access` for the second item if the first item is not present.

Read the code again. Most directories in a project do not contain any of the configuration files that are probed for. The only way to know that is to probe for all of them. Which means every new file type that is supported requires a new probe. Even in a directory where one of the supported files is present, every higher-precedence file needs to be checked before it can be used.

(Yes, I'm aware there is another approach here; it has its own drawbacks.)

> You are of course welcome to prioritize things however you want in your own code, but "prioritization" isn't applicable here. It's preference. You can also enforce your own preference in your own code, but that should be the reason given, not saying it's "redundant" when you already do that same redundant behavior.

I have a limited amount of personal time I can spend on OSS projects. And I have more things I would like to do than I have time to do them, so it is necessary for me to prioritize the things I spend time on. That's life. Of course, my personal preferences end up getting expressed in the personal projects I do. That's the privilege of them being my projects.

I'm fine if you don't like my approach - there are many other projects you might be happier using.
